Add cookie sharing projects

guardrex opened the pull request:
@HaoK

RE: Hosting cookie sharing sample projects in the AuthSamples repo. https://github.com/aspnet/Docs/pull/4981

* Fixed up the naming, including namespaces
* Fixed up the info in README; used table format to show projects+descriptions
* Confirmed that they all run locally
* Add a gitignore entry for VSC
* Hey! ... a branch! :smile:

HaoK commented on samples/CookieSharing/CookieSharing.WithIdentity.Core/Program.cs:
```diff
@@ -0,0 +1,19 @@
+﻿using Microsoft.AspNetCore;
+using Microsoft.AspNetCore.Hosting;
+
+namespace CookieSharingWithIdentityCore
```
Typically we just match the namespace with the folder, so this would be CookieSharing.WithIdentity.Core for example.  I think the nested folder is a bit redundant at this point if all 3 apps have CookieSharing in the name anyways, I'd just have all 3 live in the root like the other samples.

HaoK replied:
Actually the folder lets you have a readme, so lets keep the folder for CookieSharing

guardrex replied:
Except that I'd be linking them as a single "sample" of three projects. If we go with them in the root, then wouldn't I need to link them individually. Also, I have a concern about that *KeyRing* folder.

HaoK replied:
Either way works, its not a huge deal linking them since they all start with CookieSharing, you could include the same README.md in each app for example, whichever you prefer is fine in this case...

guardrex replied:
Let's stick with one folder for all. I'd like to have the topic have the one link ("View sample code") that lands in this folder, and the README will pop up when they arrive.

HaoK commented on samples/CookieSharing/CookieSharing.WithIdentity.NETFramework/CookieSharing.WithIdentity.NETFramework.sln:
```diff
@@ -0,0 +1,25 @@
+﻿
```
Looks like you have an extra nested directory in here, can you collapse these?

guardrex replied:
They way that had that the VS solution file was in the outer directory and the project file (and the rest of the project's assets) were in the inner folder:

* .vs
* CookieSharing.WithIdentity.NETFramework
   * CookieSharing.WithIdentity.NETFramework.csproj
   * The rest of the project's files
* packages
* CookieSharing.WithIdentity.NETFramework.sln

HaoK replied:
Hrm I guess this is a weird frankenstein situation. is there no way to just use the single AuthSamples.sln and add all projects to that?  And run them each via the same sln?

guardrex replied:
Frankenstein! Yes, we explicitly make our Core samps xplat and IDE-agnostic. ..... well ... truly, we have work to do on that point, but that's the goal.

guardrex replied:
That's what VS does when newing up a proj/sln. I have one here that I did new up a while back, and it did that on that project, too. It likes to land the sln file outside of the project's folder. I presume this is so when projects are added to the solution, the one sln file (outside of the project folders) manages them all.

![capture](https://user-images.githubusercontent.com/1622880/35076820-dfe2a7f0-fbbf-11e7-8f5b-9ad2f8d90349.PNG)

HaoK replied:
Yes I understand that, but we already have a top level AuthSamples.sln, so can we use that instead of creating a new standalone sln.  That should let you arrange it like the other projects with a single folder.

guardrex replied:
Oh, I see what you mean. Not used to the setup, since we're actively getting the last of the VS projects out of the doc samples. I'll fix it up.

HaoK replied:
Yeah repos typically have one top level sln file which include all of the projects.

HaoK replied:
AuthSamples is a bit unique since its meant as a samples repo but its also part of Universe and the CI build so it also needs to behave like a real library repo.

HaoK commented on samples/CookieSharing/README.md:
```diff
@@ -0,0 +1,11 @@
+| CookieSharing.Core                      | ASP.NET Core 2.0 Razor Pages app without using ASP.NET Core Identity |
+| CookieSharing.WithIdentity.Core         | ASP.NET Core 2.0 MVC app with ASP.NET Core Identity |
+| CookieSharing.WithIdentity.NETFramework | ASP.NET Framework 4.6.1 MVC app with ASP.NET Identity |
+
```
Maybe add a brief set of steps to demonstrate things are working

- Run the WithIdentity app, register a user, verify you are logged in
- Run the Core app, demonstrate that you are still logged in
- Run the old .NetIdentity app, demonstrate that you are still logged in

HaoK replied:
I'm assuming everything has already been setup/wired up so they'd just have to run and go to the right url with the same brower session

HaoK replied:
I'm not sure if we already have a topic for identity database interop as well, but your two samples here are perfect to demonstrate how you can use the same identity database between core + .NETFramework here as well.

guardrex replied:
Correct

guardrex commented:
@HaoK 

* Fixed namespaces and reconfirmed that the apps run and still share cookies.
* Added instructions to the README.

What about the recent security announcements updates?? Should this be added to our issue to address the needed updates, or should I do them now? I'd like to add them to the issue so that the author focusing on that task (and it might be me) can address them all at once.

[Samples that require attention for Microsoft Security Advisories aspnet/Docs \#5144](https://github.com/aspnet/Docs/issues/5144)

HaoK commented:
I'll defer to @blowdart about the recent security announcements, I'm not sure if we have done a sweep through our repos for updating everywhere yet

blowdart commented:
Samples should be up to date. So yea, the code needs updating.

HaoK commented on samples/CookieSharing/README.md:
```diff
@@ -8,4 +8,12 @@ The sample illustrates cookie sharing across three apps that use cookie authenti
 | CookieSharing.WithIdentity.NETFramework | ASP.NET Framework 4.6.1 MVC app with ASP.NET Identity |
 
+Instructions:
+
```
So isn't it a bit wonky that you can create the same user in all 3 apps?  I'd imagine this sample would be more realistic if all 3 apps were sharing the same database, maybe you can only create users in the identity apps, but its a bit weird for the example flow to ask you to create the same user in multiple apps.  if its too involved to get the databases shared in this PR then just file an issue for the future, and I'd at least drop the explicit steps to create duplicate users, and just demonstrate that you can create the user in one app, and use the cookie in the other apps without having them sign out/recreate users.

guardrex replied:
I agree, but it's in-memory dB as far as the Core projects are concerned. That's why it turned out that way. We do all (or most ... haven't surveyed every one) of the doc samples with Identity using the in-memory dB approach, and IIRC it's a standing order to do so for all future samples. I think the goal is to simplify sample use, since most of them have little to do with the actual underlying dB and don't really want to push one brand. It just makes our samples super-easy to setup and run xplat and xIDE.

HaoK replied:
That's fine for stand alone samples, but this repo is meant for demonstrate more realistic cookie/identity sharing.  We should just use an app.db or other EF shared database and demonstrate how to share identity db's between core and old as well.  We can put a caveat/commented out code for what modifications (if any) are needed for xplat (how to switch to in memory)

HaoK replied:
But I'm basically trying to shoehorn two topics into these 3 apps.

Cookie sharing is what you've currently focused on.
Identity db sharing involves cookie sharing and db sharing, which is where I'm going with this.

HaoK replied:
If you think its better to keep these two separate.

I would just do CookieSharing without identity.  CookieSharing.App1, CookieSharing.App2

And move the IdentityCore + IdentityNetFramework into its own thing Identity.Interop.Core, Identity.Interop.NetFramework

guardrex replied:
It sounds fine to me ... and easier to explain in sample instructions the way you propose to go.

What I suggest tho is perhaps we:
* Merge the current sample into the docs repo as we were originally going to do
* Open a new docs issue to refactor that topic (and kill the samp over there) when the new sample setup is ready to link over here.

The reason I float this is that the PR topic is ready and the PR sample works. The current topic is wrong ... the live sample is broken. This way, we get live with something that at least works now. It also closes a priority issue for us, and this work over here can take as long as it needs.

HaoK replied:
If we are agreed on the long term final results, I'm happy with whatever works best for you in the short term :)

guardrex replied:
The last ? is where do you need me to leave this for you to pick it up. Should I go ahead with that last item: Get the netfx project into one folder, drop it's solution file, add the netfx project to the AuthSamples solution file. I guess leave the naming alone for now.

... or do you want me to just stop now?

HaoK replied:
Yeah don't worry about the naming, if you can just put everything into the existing sln and make sure it all still runs/works, that'd be great

guardrex replied:
kk ... will do.

HaoK commented:
@blowdart someone already updated the AuthSamples?

blowdart commented:
Did they? Oh good :)

guardrex commented:
I checked, they aren't updated.

HaoK commented:
@blowdart that was a question that I assumed the answer was no since no one made any changes to this repo recently :)

guardrex commented:
*Copied from https://github.com/aspnet/Docs/issues/5238#issuecomment-373896559*

A consistent issue I see people copying from the current samples is that they're always creating local data protection providers rather than configuring the central service.
https://stackoverflow.com/a/49329480/2588374
This sample would be better:
https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/configuration/overview?tabs=aspnetcore2x#persistkeystofilesystem

HaoK commented:
@guardrex finally getting to trying to get this added to AuthSamples.  I fixed up the Core projects and have them running and cookie sharing works as expected...

I think I'll just get the Core samples in first and leave the older .netframework app out for now since its a weird hybrid where its using some core packages and some not, so I'm not sure ideally how we want to manifest these.  Maybe that belongs in the AspNetIdentity repo and we can just link to that, and hardcode some versions over there rather than trying to use latest updated 2.x packages.  

Thoughts @Tratcher ?

guardrex commented:
I'm interested in the Core sample, since I'd like to link them into the [Cookie sharing topic](https://docs.microsoft.com/aspnet/core/security/data-protection/compatibility/cookie-sharing) and drop the sample we're keeping for it in the docs repo. I can get to it when I'm done with a small group of 2.1 issues (a couple of weeks). Please ping me or close this issue to signal me that everything is ready over here.

Tratcher commented:
Starting with Core only is fine. I don't see why we can't add the Katana ones here later though.

muratg commented:
@guardrex is this PR still relevant? Does it need to get updated?

@HaoK could you do the needful? Let's either move it forward or get it closed. Thanks.

guardrex commented:
I believe that @HaoK moved the action (if not actual code) over to #45. He'll confirm; but I don't need this open for anything on the docs side. Either of you feel free to close this.

Our docs goal is to use the samples that @HaoK sets up over here for our doc. We won't maintain a sample in the docs repo for the topic. That work will be tracked by https://github.com/aspnet/Docs/issues/5238.

guardrex commented:
We can close this. We're going to address the issues involved on the docs side of things.
